htmd/charge/esp.py

```diff
@@ -3,30 +3,175 @@
+           [ 1.01384586,  1.05265777,  1.92898618],
+           [-0.10072499,  1.90416201,  0.43233528],
+           ...,
+           [-0.77827968, -2.58209328,  1.06180501],
+           [ 0.14494011, -0.39823904, -2.07498691],
+           [-1.50372053, -0.32872076,  1.19264375]])
+    >>> grid.writeXYZ('H2O_default.xyz')
+
+    >>> np.random.seed(20181113)
+    >>> grid = MoleculeGrid(mol, shell_factors=(1, 2), density=50)
+    >>> len(grid.getPoints())
+    4345
+    >>> grid.getPoints() # doctest: +NORMALIZE_WHITESPACE
+    array([[ 1.63974354,  1.22579616, -0.64424237],
+           [ 0.34472016,  0.77105343,  1.26543536],
+           [ 0.85159681, -0.45357728,  1.46317876],
+           ...,
+           [ 1.35323883, -3.08583052, -0.17044048],
+           [-0.80667116,  0.64144373, -0.99312377],
+           [ 2.45040501, -2.23107494, -0.48419202]])
+    >>> grid.writeXYZ('H2O_1_2__50.xyz')
+    """
+    def __init__(self, molecule, shell_factors=(1.4, 1.6, 1.8, 2.0), density=25):
+
+        self._molecule = molecule
+        if not isinstance(self._molecule, Molecule):
+            raise TypeError('"molecule" has to be instance of {}'.format(Molecule))
+        if self._molecule.numFrames != 1:
+            raise ValueError('"molecule" can have just one frame, but it has {}'.format(self._molecule.numFrames))
+
+        self._shell_factors = list(map(float, shell_factors))
+        for factor in self._shell_factors:
+            if factor <= 0:
+                raise ValueError('The elements of "shell_factors" have to be positive, but get {}'.format(factor))
+
+        self._density = density
+        if self._density <= 0:
+            raise ValueError('"density" has to be positive, but get {}'.format(self._density))
+
+        self._points = self._generatePoints()
+        self._points = self._filterPoints(self._points)
+
+    def _generatePoints(self):
+
+        all_points = []
+
+        for element, coord in zip(self._molecule.element, self._molecule.coords[:, :, 0]):
+            vdw_radius = radiusByElement(element)
+            for factor in self._shell_factors:
+
+                # Compute the number of point for each shell
+                radius = factor * vdw_radius
+                area = 4/3 * np.pi * radius**2
+                num_points = int(self._density * area)
+
+                # Generate points
+                points = radius * randomPointsOnSphere(num_points) + coord
+                all_points.append(points)
+
+        return np.concatenate(all_points)
+
+    def _filterPoints(self, points):
+
+        # Compute distance threshold for each atom
+        thresholds = [radiusByElement(element) for element in self._molecule.element]
+        thresholds = min(self._shell_factors) * np.array(thresholds)
+
+        # Detect the points further away for each atom than its threshold
+        distances = cdist(points, self._molecule.coords[:, :, 0])
+        is_valid = np.all(distances > thresholds, axis=1)
```

Comment:
Isn't this risky? When you distribute points on the first shell with some numerical noise you might end up removing those points. Maybe you should add a small delta to the thresholds?

Reply:
Good point! I need to fix this.